rfd/0061-tsh-aliases.md

```diff
@@ -0,0 +1,165 @@
+command's arguments.
+
+### git aliases
+
+Git aliases are defined in the repo-specific `.gitconfig` (or `.git/config`)
+or global `/etc/gitconfig` file, and are much closer conceptually to what we
+want to accomplish as they allow to define custom `git` subcommands:
+
+```ini
+[alias]
+co = checkout
+ls = log --oneline
+```
+
+Git aliases also support calling non-`git` commands (using `!` notation) and
+refer to the command's arguments (via `$@`, `$1`, etc.) which allows to build
+powerful new commands.
+
+For example, the following alias will make `git top` to show 15 most recently
+updated local branches:
+
+```ini
+[alias]
+top = "!bash -c 'git branch --sort=-committerdate | head -15'"
+```
+
+## Configuration and examples
+
+Users will define aliases in their `tsh` configuration file which is kept
+in `$TELEPORT_HOME/config/config.yaml` using the following syntax:
+
+```yaml
+aliases:
+    "<alias>": "<command>"
+```
+
+The `<alias>` can only be a top-level subcommand. In other words, users can
+define a new `tsh mycommand` alias but cannot define `tsh my command` command.
+
+A few notes regarding the `<command>`:
+
+- Unlike `git` aliases explored above, proposal is to always require the binary
+  name instead of defaulting to `tsh` and requiring special syntax otherwise.
+- The `<command>` will be fed to `exec.Command` for execution. To chain multiple
+  commands, use pipes, etc. users will explicitly use `bash -c`.
+- The `<command>` will require `$@`, `$1`, etc. to reference the arguments
+  provided to the alias command.
+
+### Examples
+
+Starting with a very simple example, suppose you're tired of typing `tsh login`
+every morning and want to alias it to something shorter e.g. `tsh l`:
+
+```yaml
+aliases:
+    "l": "tsh login $@"
+```
+
+As a more practical example, the following alias will make `tsh login` default
+to the specific auth connector and username:
+
+```yaml
+aliases:
+    "login": "tsh login --auth=local --user=alice $@"
+```
+
+Similarly, as discussed in the search-based access requests RFD, some users
+may want `tsh ssh` to default to the mode in which it will auto-request access
+to the node upon encountering an access denied error:
+
+```yaml
+aliases:
+    "ssh": "tsh ssh -P $@"
+```
+
+Note: command arguments will be resolved prior to invoking the alias. So in the
+example above, when a user runs `tsh ssh root@node1`, the alias command executed
+will be `tsh ssh -P root@node1`.
+
+### Environment variables
+
+Each alias invocation will set a few environment variables that can be consumed
+within the alias.
+
+To prevent infinite recursion, `TSH_ALIAS` variable will be set indicating the
+command is invoked as an alias. If it's detected, `tsh` won't attempt to expand
+the same command again. For example, `TSH_ALIAS=login`. This will prevent other
+`login` commands from being expanded but will allow using aliases in other
+aliases (see below for an example).
```

Comment:
Is `TSH_ALIAS` a single value? What should be the behaviour if chaining more than 2 aliases like `tsh alias-1` -> `tsh alias-2` -> `tsh alias-3`?

Also just a personal preference, I feel more like `TSH_NO_ALIAS=login`. Others may disagree also not a big deal.

Reply:
Reason I named it `TSH_ALIAS` is to indicate that this is already running as an alias but that's just an implementation detail I feel, I don't have a strong preference towards one or the other.